### PR TITLE
Integrate cudagraph to autotuning 

### PR DESCRIPTION
**Pull request by scxiao**
This change is from the upstream PR https://github.com/openai/triton/pull/3306. Integrate the changes to this fork for more correct tuning results.

### CODE REVIEW

**Comment by zhanglx13** [comment]
Does cudagraph work on AMD GPUs?

**Comment by scxiao** [comment]
> Does cudagraph work on AMD GPUs?

it calls hipgraph in pytorch.

**Comment by zhanglx13** [comment]
So have you tried this cuda/hip_graph version of do_bench(). Is there any difference?

**Comment by scxiao** [comment]
> So have you tried this cuda/hip_graph version of do_bench(). Is there any difference?

The overhead is much smaller than using `cuda.event`. For example, for the same kernel, time measured from `cuda.event` is:
`[0.1031, 0.1054, 0.0689, 0.1085, 0.0767, 0.1069, 0.1039, 0.1060, 0.1031,0.1013, 0.1041, 0.1115]`
time from cuda_graph is:
`[0.0239, 0.0239, 0.0239, 0.0238, 0.0242, 0.0240, 0.0234, 0.0240, 0.0239, 0.0237]`. 

and time variance is smaller.

**Comment by zhanglx13** [comment]
Were cuda.event and cuda_graph measuring the same application? What is the time from rocprof?

**Comment by scxiao** [comment]
> Were cuda.event and cuda_graph measuring the same application? What is the time from rocprof?

cuda_graph does not measure time, it is to reduce kernel launch overhead if there are many back-to-back kernel launches. Here we use cuda.event to measure time of multiple kernel execution in the cuda_graph. The code for time measurement is at:
https://github.com/ROCm/triton/blob/bcde44f119b37fe438040c78913fc6455db5df26/python/triton/testing.py#L69-L77.

File: python/triton/runtime/autotuner.py [hunk 1]
```diff
@@ -154,8 +160,6 @@ def run(self, *args, **kwargs):
                 self.cache[key] = builtins.min(timings, key=timings.get)
                 self.pre_hook(args, reset_only=True)
                 self.configs_timings = timings
-                if self.verbose:
```
Comment (zhanglx13) [inline python/triton/runtime/autotuner.py:157]: Why do we need to remove verbose here?
Is there another way to print the best config?

Comment (scxiao) [reply python/triton/runtime/autotuner.py:157]: We have the function kernel.get_best_config() now. It seems like we did not even use the `verbose=true` at all

Comment (zhanglx13) [reply python/triton/runtime/autotuner.py:157]: I'd suggest leave `verbose=true` functionality there. I use it all the time to print best config.
And `get_best_config` can be mis-used when autotune is not present.

Comment (scxiao) [reply python/triton/runtime/autotuner.py]: > I'd suggest leave `verbose=true` functionality there. I use it all the time to print best config. And `get_best_config` can be mis-used when autotune is not present.

OK, you are using that. I just reverted the changes.

**Review by zhanglx13 — APPROVED** [review]

**Pull request merged**